src/math/lp/nla_core.cpp

```diff
@@ -1809,18 +1838,230 @@ bool core::improve_bounds() {
     }
     return bounds_improved;
 }
-    
-void core::propagate() {
-    // disable for now
-    return;
-   // propagate linear monomials
-    m_lemmas.reset();
- 
-    m_monomial_bounds.unit_propagate();
-   
+
+bool core::is_linear(const svector<lpvar>& m, lpvar& zero_var, lpvar& non_fixed) {
+    zero_var = non_fixed = null_lpvar;
+    unsigned n_of_non_fixed = 0;
+    for (lpvar v : m) {
+        if (!var_is_fixed(v)) {
+            n_of_non_fixed++;
+            non_fixed = v;
+            continue;
+        }
+        const auto& b = get_lower_bound(v);
+        if (b.is_zero()) {
+            zero_var = v;
+            return true;
+        }
+    }
+    return n_of_non_fixed <= 1;
+
+}
+
+void core::add_lower_bound_monic(lpvar j, const lp::mpq& v, bool is_strict, std::function<u_dependency*()> explain_dep) {
+    TRACE("add_bound", lra.print_column_info(j, tout) << std::endl;);
+    j = lra.column_to_reported_index(j);
+    unsigned k;
+    if (!m_improved_lower_bounds.find(j, k)) {
+        m_improved_lower_bounds.insert(j, static_cast<unsigned>(m_implied_bounds.size()));
+        m_implied_bounds.push_back(lp::implied_bound(v, j, true, is_strict, explain_dep));
+    }
+    else {
+        auto& found_bound = m_implied_bounds[k];
+        if (v > found_bound.m_bound || (v == found_bound.m_bound && !found_bound.m_strict && is_strict)) {
+            found_bound = lp::implied_bound(v, j, true, is_strict, explain_dep);
+            TRACE("add_bound", lra.print_implied_bound(found_bound, tout););
+        }
+    }
 }
 
+    void core::add_upper_bound_monic(lpvar j, const lp::mpq& bound_val, bool is_strict, std::function<u_dependency*()> explain_dep) {
+        j = lra.column_to_reported_index(j);
+        unsigned k;
+        if (!m_improved_upper_bounds.find(j, k)) {
+            m_improved_upper_bounds.insert(j, static_cast<unsigned>(m_implied_bounds.size()));
+            m_implied_bounds.push_back(lp::implied_bound(bound_val, j, false, is_strict, explain_dep));
+        }
+        else {
+            auto& found_bound = m_implied_bounds[k];
+            if (bound_val > found_bound.m_bound || (bound_val == found_bound.m_bound && !found_bound.m_strict && is_strict)) {
+                found_bound = lp::implied_bound(bound_val, j, false, is_strict, explain_dep);
+                TRACE("add_bound", lra.print_implied_bound(found_bound, tout););
+            }
+        }
+    }
+
+    bool core::upper_bound_is_available(unsigned j) const {
+        switch (get_column_type(j)) {
+        case lp::column_type::fixed:
+        case lp::column_type::boxed:
+        case lp::column_type::upper_bound:
+            return true;
+        default:
+            return false;
+        }
+    }
+    
+    bool core::lower_bound_is_available(unsigned j) const {
```

Comment:
this one now has two implementations.
Neither appear to be the right place, which would be lar_solver since bounds are maintained there. This is not specific to non-linear

Reply:
how are these functions related to column_has_upper/lower_bound in lp_core_solver_base.h, and re-exposed from lar_solver?